cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,355 @@
+}
+
+func NewTransactionWriter(db db.SessionInterface, stmtCache *sq.StmtCache, passphrase string) TransactionWriter {
+	return &transactionHandler{db: db, stmtCache: stmtCache, passphrase: passphrase}
+}
+
+func (txn *transactionHandler) InsertTransactions(lcm xdr.LedgerCloseMeta) error {
+	start, mid := time.Now(), time.Now()
+	txCount := lcm.CountTransactions()
+	L := log.
+		WithField("ledger_seq", lcm.LedgerSequence()).
+		WithField("tx_count", txCount)
+
+	defer func() {
+		if txn.ingestMetric != nil {
+			txn.ingestMetric.Observe(time.Since(start).Seconds())
+			txn.insertMetric.Observe(time.Since(mid).Seconds())
+			txn.countMetric.Observe(float64(txCount))
+		}
+	}()
+
+	if txn.stmtCache == nil {
+		return errors.New("TransactionWriter incorrectly initialized without stmtCache")
+	} else if txCount == 0 {
+		return nil
+	}
+
+	reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(txn.passphrase, lcm)
+	if err != nil {
+		return errors.Wrapf(err,
+			"failed to open transaction reader for ledger %d",
+			lcm.LedgerSequence())
+	}
+
+	transactions := make(map[xdr.Hash]ingest.LedgerTransaction, txCount)
+	for i := 0; i < txCount; i++ {
+		tx, err := reader.Read()
+		if err != nil {
+			return errors.Wrapf(err, "failed reading tx %d", i)
+		}
+
+		// For fee-bump transactions, we store lookup entries for both the outer
+		// and inner hashes.
+		if tx.Envelope.IsFeeBump() {
+			transactions[tx.Result.InnerHash()] = tx
+		}
+		transactions[tx.Result.TransactionHash] = tx
+	}
+
+	mid = time.Now()
+	query := sq.Insert(transactionTableName).
+		Columns("hash", "ledger_sequence", "application_order")
+	for hash, tx := range transactions {
+		query = query.Values(hash[:], lcm.LedgerSequence(), tx.Index)
+	}
+	_, err = query.RunWith(txn.stmtCache).Exec()
+
+	L.WithError(err).
+		WithField("duration", time.Since(start)).
+		WithField("sql_duration", time.Since(mid)).
+		Infof("Ingested %d transaction lookups", len(transactions))
+
+	return err
+}
+
+func (txn *transactionHandler) RegisterMetrics(ingest, insert, count prometheus.Observer) {
+	txn.ingestMetric = ingest
+	txn.insertMetric = insert
+	txn.countMetric = count
+}
+
+// trimTransactions removes all transactions which fall outside the ledger retention window.
+func (txn *transactionHandler) trimTransactions(latestLedgerSeq uint32, retentionWindow uint32) error {
+	if latestLedgerSeq+1 <= retentionWindow {
+		return nil
+	}
+
+	cutoff := latestLedgerSeq + 1 - retentionWindow
+	_, err := sq.StatementBuilder.
+		RunWith(txn.stmtCache).
+		Delete(transactionTableName).
+		Where(sq.Lt{"ledger_sequence": cutoff}).
+		Exec()
+	return err
+}
+
+// NewTx creates a read-only SQL transaction on a cloned database session. You
+// MUST call `.Done()` on the resulting reader if there are no errors.
+func (txn *transactionHandler) NewTx(ctx context.Context) (TransactionReader, error) {
+	sesh := txn.db.Clone()
+	if err := sesh.BeginTx(ctx, &sql.TxOptions{ReadOnly: true}); err != nil {
```

Comment:
I don't think it's necessary to start a transaction for `GetTransaction(hash xdr.Hash)` and `GetLedgerRange()`. The reason we start a repeatable read transaction for the `LedgerEntryReader` is because we may need to query multiple ledger entries for a single simulate transaction or get ledger entries request and, in that scenario, we want to ensure all the ledger entries are consistent with the same ledger.

Transactions, unlike ledger entries, are immutable so we don't need to have a snapshot isolation level for those queries. Furthermore, we are only querying a single transaction in the `getTransaction` endpoint

Reply:
Agreed! I opted for the tx-based approach to conform to the other stuff (I asked about it [on Slack](https://stellarfoundation.slack.com/archives/G01FUHTUQ8Z/p1713217110553329), too), but it felt excessive. I've refactored in d74e30d with this simplification!